Flush the WS frames before shutting down the socket

Ashod opened the pull request:
- wsd: log when simulating socket error to help troubleshoot
- wsd: flush the WS buffered data when shutting down

Ashod commented:
@tml1024 I expect the flushing to be neutral, or helpful, for the Mobile case (and not to cause any harm otherwise).

Ideally, Mobile and non-Mobile cases would be identical except for the buildFrame call in the non-Mobile case, so I tried to go for the most common approach.

But please do take a look and let me know if you think this can have any negative impact on the Mobile case. If so, I'll move this under !MOBILEAPP.

Thanks.

mmeeks reviewed (changes requested):
Not happy with this. Magic number of '5' ? not acceptable in a live code path.
What is the problem when shutting down? Every socket that has un-sent data must be in a SocketPoll. If that is not so there is a serious bug. The long commit message didn't make it clear where we had a socket that was not in a socket poll.

The EAGAIN emulation is indeed there to catch exactly this =) but the fix is exactly not this. Blocking the whole thread for 50ms and re-writing is ... amazing.

We need to work out why the closing socket is not in a proper poll, and fix that.

More details on that ?

mmeeks reviewed (commented):
Printing here in the case that we simulate an error seems very sensible; happy to get that one in =)

Ashod commented:
> We need to work out why the closing socket is not in a proper poll, and fix that.
> 
> More details on that ?

It's a special case, but not an unusual one.

It's not an issue of not being in a poll, rather this happens *after* we exit the poll and need to close the socket. That last frame that we write (the WS Close Frame) does not get sent if we get EAGAIN at that last write, when we have already broken poll. Of course any buffered data is also dropped (without this flushing logic I added).

It happens when the document finishes (for any reason) and DocumentBroker needs to cleanup (upload to storage, notify clients, etc.). In that case, the Close Frame needs to be properly flushed (it's generated in WS shutdown). We could of course do this elsewhere, if not in the socket shutdown.

> Blocking the whole thread for 50ms and re-writing is ... amazing.

Normally we retry EAGAIN in a tight loop. We can do that here too, but I don't think it's very helpful if the socket is backed-up and we only can do the retry so many times before giving up (we could replace the magic retry count with a magic timeout, but same difference). Perhaps 50ms is too long. But as I said above, there is no polling here. If we return, the socket is closed and the buffer is not flushed. That's the hard fact. 

So we need some retrying flush of socket data. Somewhere. Somehow.

Hope it's a bit clearer now.

P.S. This is the root cause of a number of random test failures and also the root cause for the sometimes-missing "server recycling" message in the browser: the last bits of communication with the client are lost when we get EAGAIN on the final write in the socket shutdown.

mmeeks commented:
Ok - so that is clearer; but we already have rather cleaner code for this in the DocumentBroker.cpp surely:

    // Flush socket data first.
    constexpr auto flushTimeoutMicroS = std::chrono::microseconds(POLL_TIMEOUT_MICRO_S * 2); // ~1000ms
    LOG_INF("Flushing socket for doc ["
            << _docKey << "] for " << flushTimeoutMicroS << ". stop: " << _stop
            << ", continuePolling: " << _poll->continuePolling()
            << ", ShutdownRequestFlag: " << SigUtil::getShutdownRequestFlag()
            << ", TerminationFlag: " << SigUtil::getTerminationFlag()
            << ". Terminating child with reason: [" << _closeReason << "].");
    const auto flushStartTime = std::chrono::steady_clock::now();
    while (_poll->getSocketCount())

Are we missing something similar in the Kit ? should we be blocking any input read/processing during such occasions or something or ? ...

Should we move that loop to the SocketPoll class with some 'flushQueuedWrites' method or somethign to share that ? no idea - but really, really not a fan of adding extreme corner-case things with sleeps in them to the middle of the websocket code =)

Ashod commented:
> Are we missing something similar in the Kit ? should we be blocking any input read/processing during such occasions or something or ? ...

Possibly. Will double check.

> Should we move that loop to the SocketPoll class with some 'flushQueuedWrites' method or somethign to share that ? no idea - but really, really not a fan of adding extreme corner-case things with sleeps in them to the middle of the websocket code =)

Agreed. Will rework to avoid sleeping.

mmeeks commented:
Thanks ! =)

Ashod commented:
@mmeeks Turns out the edge case I found was known and recognized in a FIXME in the shutdown logic. The fix is non-trivial, but we should do it eventually (it's not too complex either, just need to do async-shutdown, which we currently do, for the case in question as well, and let poll do its magic).

So, for now, I documented my finds properly, I simply do 1 retry when writeOutgoingData doesn't consume the full buffer, and log the fact. I also unify the handling of the 'flush' case with mobile.

This should give us a reasonable safety net when we hit EAGAIN and the like upon closing, until we do proper async-shutdown with polling.

For further details, see the commit log (updated) and the comment and stack trace in the code itself.

Thanks!

Ashod commented:
> Looks fine - will avoid my EAGAIN simulation logic for now; thanks Ash =)

Thanks!
